### PR TITLE
Result_Engine, Graphics_Engine: Add DisplayMeshResults, MapResults, and methods for GradientOptions

### LINKED ISSUE #2318: Results_Engine: Migrate MapResults and DisplayMeshResults to Results_Engine
#### Description:

Migrate MapResults and DisplayMeshResults to Results_Engine from StructuralEngineering_Toolkit, while also making them more generally applicable to different types of results.

### LINKED ISSUE #2319: Graphics_Engine: Migrate Gradient methods to Graphics_Engine
#### Description:

Migrate Gradient methods currently located in StructuralEngineering_Toolkit to Graphics_Engine.

### PR DESCRIPTION
**Pull request by IsakLarbornBH**
### NOTE: Depends on 

* BHoM/BHoM#1232
   
### Issues addressed by this PR

Closes #2318 
Closes #2319 

**Result_Engine:**
Adds DisplayMeshResults, which displays IMeshResults on IMeshes, and MapResults, which maps IResults to IBHoMObjects using common identifiers such as object IDs. 
**Graphics_Engine:**
Adds methods for dealing with GradientOptions objects from BHoM/BHoM#1232.

All methods added have been adapted from methods that were previously in the StructuralEngineering_Toolkit.

### Test files

A messy but functional test file can be found here. This will allow you to push a simple test model to Robot, execute an analysis and then pull the results. You can of course use it with your own Robot file and just pull the results from that, if you prefer.
[MeshResult Test.zip](https://github.com/BHoM/BHoM_Engine/files/6067573/MeshResult.Test.zip)

If you don't want to use Robot, here is a simpler test file with some internalised FEMeshes and results:
[MeshResult Internalised.zip](https://github.com/BHoM/BHoM_Engine/files/6067809/MeshResult.Internalised.zip)

Note that both of these contain the `DisplayMeshResultsWorkaround` method which has been removed since #2367 is now merged, fixing the issue that was being worked around.

Unit tests have been added for DisplayMeshResults, which calls MapResults and all the GradientOptions methods.

### Additional comments

This pull request is one in a series that should, to minimise disruption, be merged on the same day.
* BHoM/BHoM#1232
* BHoM/BHoM_Engine#2355
* BuroHappoldEngineering/StructuralEngineering_Toolkit#419

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by FraserGreenroyd** [comment]
@IsakLarbornBH I feel this PR has become a bit mammoth looking at the commit history and the comments in the original issue regarding related PRs. It might be worth us having a catch up next sprint to go through this and get it ready so we can make sure this gets adequate reviewing resource from @IsakNaslundBh and others? :smile:

**Comment by IsakLarbornBH** [comment]
> @IsakLarbornBH I feel this PR has become a bit mammoth looking at the commit history and the comments in the original issue regarding related PRs. It might be worth us having a catch up next sprint to go through this and get it ready so we can make sure this gets adequate reviewing resource from @IsakNaslundBh and others? 😄

@FraserGreenroyd For sure, though it was always going to be a pretty big one. Although most of the commits are tiny edits, or adding and then removing temporary workarounds I had to use before I had migrated everything, so the commit history is a bit inflated and not as bad as it may seem at first glance.

Still though, definitely worth it to have a catch-up! This PR is going to take some work to review, and we should start early on that so we can be sure to get it merged in 4.2.

**Comment by rolyhudson** [comment]
Hand up here from me. 
I am keen to help get this series of PRs merged as part of a bigger (epic) foundational workflow related to representation and visualisation. 
Think we need to avoid this though:
> You must manually create a branch merging this branch and the branch referenced in #2354 in order to test this. 

I say it is better to combine this branch with #2354 and close this PR.

Also happy to be part of any discussion if deemed necessary to move this forward

**Comment by IsakLarbornBH** [comment]
> I say it is better to combine this branch with #2354 and close this PR.

I'd be more than happy to do so, though I would keep this PR and close #2354 instead, as this is the main PR.

File: Graphics_Engine/Modify/ApplyGradientOptions.cs
Comment (rolyhudson) [inline Graphics_Engine/Modify/ApplyGradientOptions.cs:29]: Are there really 3 previous versions?

File: Results_Engine/Results_Engine.csproj
Comment (rolyhudson) [inline Results_Engine/Results_Engine.csproj:43]: All the BHoM refs here should be C:\ProgramData\BHoM\Assemblies\....
And copy local set as false 
Also no specific version, culture or public key
The Structure_oM reference is correct.

File: Graphics_Engine/Modify/ApplyGradientOptions.cs
Comment (IsakLarbornBH) [reply Graphics_Engine/Modify/ApplyGradientOptions.cs:29]: Originally they were 3 separate methods, but they were made in a very particular way and essentially had to be used together in a certain order, so I merged them into one. I doubt anybody was calling them from the UI anyway, so versioning might not be necessary, but if they were it does actually make sense to version them all to this merged one since it's made in such a way as to not override manually set parameters.

Comment (rolyhudson) [reply Graphics_Engine/Modify/ApplyGradientOptions.cs:29]: OK. Yes I have seem them in the StructuralEnginnering_Toolkit. Lets see of the versioning check likes that, I've not tried it before. If not the `MessageForDeleted` in the versioning json file may be an option if they are not that widely used.

File: Results_Engine/Query/DisplayMeshResults.cs
Comment (rolyhudson) [inline Results_Engine/Query/DisplayMeshResults.cs:212]: Any good reason to keep this commented out private method? Or should it be cleaned out?

File: Results_Engine/Results_Engine.csproj
Comment (IsakLarbornBH) [reply Results_Engine/Results_Engine.csproj:45]: That should be fixed now.

File: Results_Engine/Query/DisplayMeshResults.cs
Comment (IsakLarbornBH) [reply Results_Engine/Query/DisplayMeshResults.cs:212]: This is @IsakNaslundBh who attempted to create a more optimised replacement for my calls to PropertyValue, but we never got it to work and when I read up about it later I got the impression that it wouldn't work in this case, though it was a while ago and I don't remember the details. Personally I haven't noticed the PropertyValue calls slowing this method down one bit, so I'd be happy to remove it.

**Comment by IsakLarbornBH** [comment]
@BHoMBot check core

**Comment by IsakLarbornBH** [comment]
@BHoMBot check compliance

**Comment by IsakLarbornBH** [comment]
@BHoMBot check versioning

**Comment by IsakLarbornBH** [comment]
After some rearranging and renaming of the branches, and a small code change (guessing I did something that's not compatible with older C# versions) this PR now passes the build checks. It fails the code compliance check due to the migrated Modify methods not being up to current standards, but after discussion with @FraserGreenroyd it seems we'll let them be.

File: Results_Engine/Query/MapResults.cs
Comment (rolyhudson) [inline Results_Engine/Query/MapResults.cs:124]: Why is this not using the: Engine.Reflection.Compute.RecordError("No Identifier found"); ?

Comment (IsakLarbornBH) [reply Results_Engine/Query/MapResults.cs:124]: I haven't a clue to be honest, I have not touched this piece of code myself. Possibly it's done to stop execution of the method?

File: Graphics_Engine/Modify/CenterGradientAsymmetric.cs
Comment (rolyhudson) [inline Graphics_Engine/Modify/CenterGradientAsymmetric.cs:44]: Inputs `to` and `from` should probably be `upperBound` and `lowerBound` for consistency with changes you made to `GradientOptions`

Comment (IsakLarbornBH) [reply Graphics_Engine/Modify/CenterGradientAsymmetric.cs:46]: Good point, fixed it.

**Review by rolyhudson — APPROVED** [review]
I think this is good to go now. As a non-robot user I have tested against the script in MeshResult Internalised.zip all working as expected. 

To generalise this further we should consider a wider set of analysis and result representation workflows. Objects and results for which we wish to `DisplayMeshResult` won't necessarily come from an adapter and won't have IAdapterId fragment which is used here to match results to objects.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

File: Results_Engine/Results_Engine.csproj
Comment (IsakNaslundBh) [inline Results_Engine/Results_Engine.csproj:113]: Is this reference required, or leftover from prototyping and testing? Would be good if we could get away with not having this reference in.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project-compliance

File: Results_Engine/Results_Engine.csproj
Comment (IsakLarbornBH) [reply Results_Engine/Results_Engine.csproj:113]: I think we should be able to get rid of some references, Structure_oM is required due to MeshResultSmoothingType but I don't think Structure_Engine should be. I don't personally have time to remove and test today unfortunately, so unless someone else does it earlier I'll clean up the references next week.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project-compliance

File: Results_Engine/Results_Engine.csproj
Comment (IsakNaslundBh) [reply Results_Engine/Results_Engine.csproj:113]: Yeah, keeping Structure_oM is fine, it was already in here, required by the envelope methods for BarForces (might need to move that out at one point as well, but that is for sure nothing that this PR should have to worry about!)
Was mainly concerned of any (hopefully) unnecessary engine cross dependencies, as I can see a case of the structures_Engine wanting to depend on results engine going forward.

Comment (IsakNaslundBh) [reply Results_Engine/Results_Engine.csproj:113]: Also, no rush to get this done today! I will struggle for time to fully review today anyway, just spotted it and put a comment when I saw it!

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Comment by IsakLarbornBH** [comment]
@BHoMBot check required

File: Results_Engine/Results_Engine.csproj
Comment (IsakLarbornBH) [reply Results_Engine/Results_Engine.csproj]: The Structure_Engine references and a couple other unused ones added by this have been removed.

File: Results_Engine/Query/DisplayMeshResults.cs
Comment (alelom) [reply Results_Engine/Query/DisplayMeshResults.cs:215]: I think you can delete this. If the performance becomes a problem, a separate PR will target this.

File: Results_Engine/Query/MapResults.cs
Comment (alelom) [reply Results_Engine/Query/MapResults.cs:123]: As you're moving over code from another place, I think the missing use of RecordError here should not be part of this review. (Let's leave the code as it was)

**Review by alelom — APPROVED** [review]
Code LGTM. I have tested with the provided test file and I'm happy with the result.

<details>

image

</details>

**Review by rolyhudson — CHANGES_REQUESTED** [review]
I'm happy with this - good to see @alelom has reviewed with a structural point of view. If you could remove the commented out unused method this is fine.

**Comment by IsakLarbornBH** [comment]
> I'm happy with this - good to see @alelom has reviewed with a structural point of view. If you could remove the commented out unused method this is fine.

@rolyhudson Already did that, should maybe have mentioned it on here as well :)

**Review by alelom — APPROVED** [review]
Happy with the changes done by @IsakLarbornBH . LGTM.

**Review by rolyhudson — APPROVED** [review]
LGTM now!

**Comment by rolyhudson** [comment]
@alelom I will leave it to you to remove the do-not-merge when you are happy.

**Comment by alelom** [comment]
Done, thanks @rolyhudson .

**Comment by peterjamesnugent** [comment]
@BHoMBot check required

**Comment by IsakLarbornBH** [comment]
@BHoMBot fix project file ref. 2640186858

**Comment by IsakLarbornBH** [comment]
@BHoMBot check required

**Comment by IsakLarbornBH** [comment]
@BHoMBot check null-handling

**Comment by IsakLarbornBH** [comment]
@peterjamesnugent Well I tried getting some null handling in there, unfortunately it still fails the check. I don't know if anybody's able to extract any useful information from the logs it produces but I'm certainly not. I can't even find any of the methods in this PR listed in the logs.

**Comment by FraserGreenroyd** [comment]
@IsakLarbornBH det låter bra till mig

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copy-right compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check dataset-compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copyright-compliance

**Comment by IsakLarbornBH** [comment]
Fixed the failing copyright compliance check, hopefully:

@BHoMBot check copyright-compliance

It passed dataset compliance though.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check dataset-compliance

**Review by IsakNaslundBh — DISMISSED** [review]
This all looks fine to me.

Checks failing is out of scope of this PR, especially after @peterjamesnugent s check.

Tested and seem to upgrade fine.

**Comment by IsakNaslundBh** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 2709053021

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready to merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Comment by IsakNaslundBh** [comment]
Unfortunately had to locally rebase and force push this as well due to github complaining about conflicts. Rebase locally went with 0 hickups and no conflicts showing up. Not sure why Github did not like it, but hopefully will be fixed now.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copyright-compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check dataset-compliance

**Review by IsakNaslundBh — APPROVED** [review]
Re-approving after testing again.

Local rebase did not throw up any conflicts, so think previous approvals are still safe as well!

**Comment by IsakNaslundBh** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 2717219160

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**
